presto-memory/src/main/java/io/prestosql/plugin/memory/MemoryMetadata.java

```diff
@@ -126,60 +124,63 @@ public synchronized void dropSchema(ConnectorSession session, String schemaName)
     @Override
     public synchronized ConnectorTableHandle getTableHandle(ConnectorSession session, SchemaTableName schemaTableName)
     {
-        Long tableId = tableIds.get(schemaTableName);
-        if (tableId == null) {
+        TableId id = tableIds.get(schemaTableName);
+        if (id == null) {
             return null;
         }
-        return tables.get(tableId);
+
+        return new MemoryTableHandle(id);
     }
 
     @Override
     public synchronized ConnectorTableMetadata getTableMetadata(ConnectorSession session, ConnectorTableHandle tableHandle)
     {
-        MemoryTableHandle memoryTableHandle = (MemoryTableHandle) tableHandle;
-        return memoryTableHandle.toTableMetadata();
+        MemoryTableHandle handle = (MemoryTableHandle) tableHandle;
+        return tables.get(handle.getId()).getMetadata();
     }
 
     @Override
     public synchronized List<SchemaTableName> listTables(ConnectorSession session, Optional<String> schemaName)
     {
         return tables.values().stream()
                 .filter(table -> schemaName.map(table.getSchemaName()::equals).orElse(true))
-                .map(MemoryTableHandle::toSchemaTableName)
+                .map(TableInfo::getSchemaTableName)
                 .collect(toList());
     }
 
     @Override
     public synchronized Map<String, ColumnHandle> getColumnHandles(ConnectorSession session, ConnectorTableHandle tableHandle)
     {
-        MemoryTableHandle memoryTableHandle = (MemoryTableHandle) tableHandle;
-        return memoryTableHandle.getColumnHandles().stream()
-                .collect(toMap(MemoryColumnHandle::getName, Function.identity()));
+        MemoryTableHandle handle = (MemoryTableHandle) tableHandle;
+        return tables.get(handle.getId())
```

Comment:
I can't remember... is it ok if this throws an NPE?

Reply:
Yeah, any RuntimeException if tableHandle is invalid is ok

Reply:
Yeah, any RuntimeException if tableHandle is invalid is ok